src/Compilers/CSharp/Test/Semantic/Semantics/CollectionLiteralTests.cs

```diff
@@ -15,290 +15,1566 @@ namespace Microsoft.CodeAnalysis.CSharp.UnitTests
+            Assert.Equal(2, collections.Length);
+            VerifyTypes(model, collections[0], "System.Collections.Generic.List<nint>", "System.Collections.Generic.List<nint>", ConversionKind.Identity);
+            VerifyTypes(model, collections[1], "System.Collections.Generic.List<nuint[]>", "System.Collections.Generic.List<nuint[]>", ConversionKind.Identity);
+        }
+
+        [Fact]
+        public void NaturalType_27()
+        {
+            string source = """
+                #nullable enable
+                class Program
+                {
+                    static object F1(object? a, object b)
+                    {
+                        var c1 = [a, b];
+                        return c1;
+                    }
+                    static object F2(object?[] a, object[] b)
+                    {
+                        var c2 = [b, a];
+                        return c2;
+                    }
+                    static void Main()
+                    {
+                        F1(1, 2).Report();
+                        F2(new object[] { 3 }, new object[] { 4 }).Report();
+                    }
+                }
+                """;
+            var comp = CreateCompilation(new[] { source, s_collectionExtensions }, options: TestOptions.ReleaseExe);
+            comp.VerifyEmitDiagnostics(
+                // 0.cs(11,22): warning CS8620: Argument of type 'object?[]' cannot be used for parameter 'item' of type 'object[]' in 'void List<object[]>.Add(object[] item)' due to differences in the nullability of reference types.
+                //         var c2 = [b, a];
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInArgument, "a").WithArguments("object?[]", "object[]", "item", "void List<object[]>.Add(object[] item)").WithLocation(11, 22));
+            CompileAndVerify(comp, expectedOutput: "[1, 2], [[4], [3]], ");
+
+            var tree = comp.SyntaxTrees[0];
+            var model = comp.GetSemanticModel(tree);
+            var collections = tree.GetRoot().DescendantNodes().OfType<CollectionCreationExpressionSyntax>().ToArray();
+            Assert.Equal(2, collections.Length);
+            VerifyTypes(model, collections[0], "System.Collections.Generic.List<System.Object>", "System.Collections.Generic.List<System.Object>", ConversionKind.Identity);
+            VerifyTypes(model, collections[1], "System.Collections.Generic.List<System.Object[]>", "System.Collections.Generic.List<System.Object[]>", ConversionKind.Identity);
+        }
+
+        [Fact]
+        public void NaturalType_28()
+        {
+            string source = """
+                class Program
+                {
+                    static void Main()
+                    {
+                        F(false);
+                    }
+                    static void F(bool b)
+                    {
+                        var x = b ? [1] : [];
```

Comment:
Consider a version of this with `[null]` in the other hand. Also consider a version where they're target-typed to `int?[]`

Reply:
Ah, I see you have _32 for the null suggestion. Consider adding one for the target-typed suggestion though.

Reply:
Added `TargetType_01` and `_02`.